sdk/scheduler/src/main/java/com/mesosphere/sdk/curator/CuratorStateStore.java

```diff
@@ -193,14 +192,45 @@ public void storeTasks(Collection<Protos.TaskInfo> tasks) throws StateStoreExcep
 
     @Override
     public void storeStatus(Protos.TaskStatus status) throws StateStoreException {
-        String taskName;
+        Optional<String> optionalTaskName = Optional.empty();
+
+        //TODO(MB): keep the commented code below: original approach for getting TaskName
+        /* We used to get TaskName from TaskID's prefix, instead of comparing TaskIDs.
+           TaskID is our unique identifier, can not change, but we can modify TaskName
+
+           Getting TaskName from prefix may be more efficient, however it prevents us from changing TaskName.
+           For further investigation, keep the commented code below:
+         */
+        /*
         try {
-            taskName = CommonTaskUtils.toTaskName(status.getTaskId());
+            optionalTaskName = Optional.of(CommonTaskUtils.toTaskName(status.getTaskId()));
         } catch (TaskException e) {
             throw new StateStoreException(Reason.LOGIC_ERROR, String.format(
                     "Failed to parse the Task Name from TaskStatus.task_id: '%s'", status), e);
         }
+        */
+
+        for (Protos.TaskInfo taskInfo : fetchTasks()) {
+            if (taskInfo.getTaskId().equals(status.getTaskId())) {
+                if (optionalTaskName.isPresent()) {
+                    logger.error("Found identical task ids in Task {} and  Task {}",
+                            optionalTaskName.get(), taskInfo.getName());
+                    throw new StateStoreException(Reason.LOGIC_ERROR, String.format(
+                            "There are more than one tasks with the same TaskStatus.task_id: %s", status));
+                }
+                    optionalTaskName = Optional.of(taskInfo.getName());
+            }
+        }
+        if (!optionalTaskName.isPresent()) {
+            throw new StateStoreException(Reason.NOT_FOUND, String.format(
+                    "Failed to find a task with matching TaskStatus.task_id: %s", status));
```

Comment:
Please don't use this notation `TaskStatus.task_id`.  Just say `TaskID`.  e.g. `Failed to find a task with TaskID: %s`